### PR TITLE
feat(embodiment): add OpenVLA-OFT OPD training

### PR DESCRIPTION
**Pull request by jzndd**
Implement OPD for LIBERO spatial with post-rollout teacher logprob computation, masked token-level actor loss, and stabilized training configuration.

### Description
This PR adds `libero_spatial_opd_openvlaoft`, an OPD training recipe for OpenVLA-OFT on LIBERO-Spatial.

Main changes:
- Add OPD advantage/loss path for embodied training.
- Compute teacher logprobs after student rollout, using saved student action tokens instead of generating teacher actions during rollout.
- Apply token-level `loss_mask` to OPD actor loss so invalid tokens after success/termination do not contribute gradients.
- Add neutral expert/teacher config helper without coupling common workers to RLT-specific modules.
- Add the LIBERO-Spatial OPD config with stabilized rollout and actor batch settings.
- Add Chinese and English documentation for the OPD example, including model weights, config fields, run command, and result table.

TODO list:
- [ ] Add a image of OPD (RLinf/misc)
- [ ] Add a CI test 
- [ ] Update README.md.

### Motivation and Context
Add an opd support.

### How has this been tested?
- Verified generated OPD docs pages render correctly.
- Verified the rising curve of success rates.

### Additional information (optional, e.g., figures and logs):
0a58547a6afabfd3f9fd32b91d1c449e

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Documentation update (Document-only update)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: docs/source-en/rst_source/examples/embodied/opd.rst [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+
+   source switch_env openvla-oft
+   bash examples/embodiment/run_embodiment.sh libero_spatial_opd_openvlaoft
+
+Logs are written under ``runner.logger.log_path`` and can be inspected with TensorBoard:
+
+.. code-block:: bash
+
+   tensorboard --logdir ./logs --port 6006
+
```
[the remark this answers is not in the document]
Comment (jzndd) [reply docs/source-en/rst_source/examples/embodied/opd.rst:153]: unnecessary

File: docs/source-zh/rst_source/examples/embodied/opd.rst [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+   source switch_env openvla-oft
+   bash examples/embodiment/run_embodiment.sh libero_spatial_opd_openvlaoft
+
+
+训练日志写入 ``runner.logger.log_path``，默认使用 TensorBoard：
+
+.. code-block:: bash
+
+   tensorboard --logdir ./logs --port 6006
+
```
[the remark this answers is not in the document]
Comment (jzndd) [reply docs/source-zh/rst_source/examples/embodied/opd.rst:149]: unnecessary

File: examples/embodiment/config/libero_spatial_opd_openvlaoft.yaml [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+
+  group_name: "RolloutGroup"
+  generation_backend: "huggingface"
+  enable_offload: True
+  pipeline_stage_num: 1
+
+  model:
+    model_path: "/mnt/project_rlinf/jzn/ckpt_path/RLinf/RLinf-OpenVLAOFT-LIBERO-130-Base-Lora"
+    precision: ${actor.model.precision}
+
```
[the remark this answers is not in the document]
Comment (jzndd) [reply examples/embodiment/config/libero_spatial_opd_openvlaoft.yaml:116]: Solved.

File: examples/embodiment/config/libero_spatial_opd_openvlaoft.yaml [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+  expert_model:
+    model_path: "/mnt/project_rlinf/jzn/ckpt_path/RLinf/RLinf-OpenVLAOFT-LIBERO-130"
+    precision: ${actor.model.precision}
+    unnorm_key: libero_130_no_noops_trajall
+
```
[the remark this answers is not in the document]
Comment (jzndd) [reply examples/embodiment/config/libero_spatial_opd_openvlaoft.yaml:121]: Solved.

File: examples/embodiment/config/libero_spatial_opd_openvlaoft.yaml [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+  training_backend: "fsdp"
+  micro_batch_size: 32
+  global_batch_size: 4096
+  seed: 1234
+  enable_offload: True
+
+  # Override the default values in model/openvla_oft
+  model:
+    model_path: "/mnt/project_rlinf/jzn/ckpt_path/RLinf/RLinf-OpenVLAOFT-LIBERO-130-Base-Lora"
+    model_type: "openvla_oft"
```
[the remark this answers is not in the document]
Comment (jzndd) [reply examples/embodiment/config/libero_spatial_opd_openvlaoft.yaml:133]: Solved.

File: rlinf/workers/actor/fsdp_actor_worker.py [hunk 6]
```diff
@@ -1269,6 +1316,77 @@ def compute_advantages_and_returns(self) -> dict[str, torch.Tensor]:
...
+
+        prev_logprobs = self.rollout_batch["prev_logprobs"]
+        time_dim, batch_dim = prev_logprobs.shape[:2]
+        flat_batch_size = time_dim * batch_dim
+
+        teacher_model_config = build_expert_model_config(self.cfg, self.cfg.actor.model)
+        teacher_model = get_model(teacher_model_config)
+        if self.cfg.runner.get("expert_ckpt_path", None):
+            teacher_model_dict = torch.load(self.cfg.runner.expert_ckpt_path)
+            teacher_model.load_state_dict(teacher_model_dict)
```
[the remark this answers is not in the document]
Comment (jzndd) [reply rlinf/workers/actor/fsdp_actor_worker.py]: Solved in latest version.

File: rlinf/workers/env/env_worker.py [hunk 7]
```diff
@@ -1365,6 +1373,8 @@ def compute_advantages_and_returns(
             "rewards": rollout_batch["rewards"],
             "dones": rollout_batch["dones"],
             "values": rollout_batch.get("prev_values", None),
+            "prev_logprobs": rollout_batch.get("prev_logprobs", None),
+            "num_action_chunks": self.cfg.actor.model.num_action_chunks,
             "gamma": self.cfg.algorithm.get("gamma", 1),
```
[the remark this answers is not in the document]
Comment (jzndd) [reply rlinf/workers/env/env_worker.py:1387]: Solved in latest version.

File: rlinf/workers/actor/fsdp_actor_worker.py [hunk 8]
```diff
@@ -1269,6 +1316,79 @@ def compute_advantages_and_returns(self) -> dict[str, torch.Tensor]:
...
+        teacher_model_config = build_expert_model_config(self.cfg, self.cfg.actor.model)
+        teacher_model = get_model(teacher_model_config)
+        if self.cfg.runner.get("expert_ckpt_path", None):
+            teacher_model_dict = torch.load(
+                self.cfg.runner.expert_ckpt_path, map_location="cpu"
+            )
+            teacher_model.load_state_dict(teacher_model_dict)
+        teacher_model.eval()
+        teacher_model.requires_grad_(False)
+        teacher_model.to(self.device)
```
Comment (thereAreDemonsNearby) [inline rlinf/workers/actor/fsdp_actor_worker.py:1359]: I want to confirm that at the point before the teacher_model is loaded onto the GPU, the actor model has not been loaded to the GPU yet, right? That way there's no risk of OOM

**Review by anHappyDog — APPROVED** [review]

File: rlinf/workers/actor/fsdp_actor_worker.py [hunk 9]
```diff
@@ -1362,9 +1495,36 @@ def run_training(self) -> None:
...
+            )
+        else:
+            actor_train_keys.update(
+                {
+                    "loss_mask",
+                    "loss_mask_sum",
+                    "prev_values",
+                }
+            )
+        train_rollout_batch = {
```
Comment (thereAreDemonsNearby) [inline rlinf/workers/actor/fsdp_actor_worker.py]: I'd like to know what the purpose of filtering the rollout_batch here is. Is it to save time on copying data to the GPU?

File: rlinf/workers/rollout/hf/huggingface_worker.py [hunk 10]
```diff
@@ -720,18 +722,27 @@ async def generate_one_epoch(self, input_channel: Channel, output_channel: Chann
...
-                bootstrap_values=self.get_bootstrap_values(
-                    env_output.get("final_obs", None)
-                ),
-                forward_inputs=(
-                    result["forward_inputs"]
-                    if self.rlt_feature_model is not None
-                    else {}
-                ),
-            )
+            if self.enable_opd:
```
Comment (thereAreDemonsNearby) [inline rlinf/workers/rollout/hf/huggingface_worker.py:726]: Can the if branch and the else branch here share the self._build_rollout_result function? I feel like the logic is largely overlapping.

File: rlinf/workers/actor/fsdp_actor_worker.py [hunk 11]
```diff
@@ -1269,6 +1316,79 @@ def compute_advantages_and_returns(self) -> dict[str, torch.Tensor]:
...
+        time_dim, batch_dim = prev_logprobs.shape[:2]
+        flat_batch_size = time_dim * batch_dim
+
+        teacher_model_config = build_expert_model_config(self.cfg, self.cfg.actor.model)
+        teacher_model = get_model(teacher_model_config)
+        if self.cfg.runner.get("expert_ckpt_path", None):
+            teacher_model_dict = torch.load(
+                self.cfg.runner.expert_ckpt_path, map_location="cpu"
+            )
+            teacher_model.load_state_dict(teacher_model_dict)
```
Comment (thereAreDemonsNearby) [inline rlinf/workers/actor/fsdp_actor_worker.py]: I don't think it's ideal to reload the state_dict every time compute_opd_teacher_logprobs is run. And calling get_model each time (which may involve model initialization and loading) is also not feasible.

**Review by thereAreDemonsNearby — COMMENTED** [review]
Personally, I think the expert model could be added in a more flexible way: by introducing a fourth component — teacher rollout (or expert rollout) — alongside env, rollout, and actor, and the code could reuse the HuggingFace inference worker. The embodied runner might need to be modified to integrate the teacher inference into the overall pipeline. This way, the teacher could benefit from RLinf's native support for features like offloading, and it would also avoid the issue of calling get_model every time we compute teacher logprobs.

File: rlinf/workers/actor/fsdp_actor_worker.py (hunk 8 above)
Comment (jzndd) [reply rlinf/workers/actor/fsdp_actor_worker.py:1359]: Solved.  I acknowledge that is not elegant. OPD now asserts that actor weights are offloaded before moving the teacher model to GPU. The cached teacher is only moved to GPU for teacher logprob computation, then immediately offloaded back to CPU.

File: rlinf/workers/actor/fsdp_actor_worker.py (hunk 9 above)
Comment (jzndd) [reply rlinf/workers/actor/fsdp_actor_worker.py]: Re-tested. I found that design is unnecessary. I reverted the OPD-specific filtering and now keep prev_values consistently.

File: rlinf/workers/rollout/hf/huggingface_worker.py (hunk 10 above)
Comment (jzndd) [reply rlinf/workers/rollout/hf/huggingface_worker.py:726]: Verified. Directly reusing `_build_rollout_result()` without additional controls breaks the GRPO run because extra forward_inputs fields have a different flattened length from the actor-side shuffle index. To avoid affecting the existing GRPO path, I kept the original branch behavior.

File: rlinf/workers/actor/fsdp_actor_worker.py (hunk 11 above)
Comment (jzndd) [reply rlinf/workers/actor/fsdp_actor_worker.py]: Handled. The teacher model is no longer rebuilt and reloaded every step. It is now lazily cached inside the actor worker: loaded on CPU once, moved to GPU only for teacher logprob computation, then offloaded back to CPU. The measured teacher-logprob time dropped clearly after the first step.

**Review by thereAreDemonsNearby — APPROVED** [review]

**Review by andylin-hao — APPROVED** [review]

**Review by guozhen1997 — APPROVED** [review]
